File: zig/triangle/triangle.zig
pub const Triangle = struct {
    a: f64,
    b: f64,
    c: f64,

    pub fn init(a: f64, b: f64, c: f64) TriangleError!Triangle {
        if (a <= 0 or b <= 0 or c <= 0) {
            return TriangleError.Invalid;
        }
        if (a + b >= c and b + c >= a and a + c >= b) {
            return Triangle{ .a = a, .b = b, .c = c };
        } else {
            return TriangleError.Invalid;
        }
    }

    pub fn isEquilateral(self: Triangle) bool {
        return self.a == self.b and self.b == self.c;
    }

    pub fn isIsosceles(self: Triangle) bool {
        return self.a == self.b or self.a == self.c or self.b == self.c;
    }

    pub fn isScalene(self: Triangle) bool {
        return self.a != self.b and self.a != self.c and self.b != self.c;
    }
};

pub const TriangleError = error{
    Invalid,
};

File: typescript/triangle/triangle.ts
export class Triangle {
  private a: number
  private b: number
  private c: number
  constructor(...sides: Array<number>) {
    this.a = sides[0]
    this.b = sides[1]
    this.c = sides[2]
  }


  get isEquilateral() {
    if (!this.isTriangle()) return false
    return this.a === this.b && this.b === this.c
  }

  get isIsosceles() {
    if (!this.isTriangle()) return false
    return this.a === this.b || this.a === this.c || this.b === this.c;
  }

  get isScalene() {
    if (!this.isTriangle()) return false;
    return this.a !== this.b && this.a !== this.c && this.b !== this.c;
  }

  isTriangle(): boolean {
    if (this.a <= 0 || this.b <= 0 || this.c <= 0) {
      return false
    }
    if (!(this.a + this.b >= this.c && this.b + this.c >= this.a && this.a + this.c >= this.b)) {
      return false
    }
    return true
  }
}
